facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+                let start = values
+                    .next()
+                    .context(format!("unexpected Range header {}", range_header))?
+                    .parse::<usize>()
+                    .context(format!("unexpected Range header {}", range_header))?;
+                let end = values
+                    .next()
+                    .context(format!("unexpected Range header {}", range_header))?
+                    .parse::<usize>()
+                    .context(format!("unexpected Range header {}", range_header))?;
+
+                if start != self.object_upload_position {
+                    return Err(anyhow!("unexpected Range header {}", range_header));
+                }
+
+                self.object_upload_position = end + 1;
+
+                // If we have a little content left over, we can't just make
+                // another request, because if there's too little of it, Google
+                // will reject it. Instead, put the portion of the chunk that we
+                // didn't manage to upload back into self.buffer so it can be
+                // handled by a subsequent call to upload_chunk.
+                self.buffer
+                    .extend_from_slice(&chunk[self.object_upload_position - start..]);
+            }
+            _ => {
+                return Err(anyhow!(
+                    "failed to upload part to GCS: {} synthetic: {}\n{:?}",
+                    http_response.status(),
+                    http_response.synthetic(),
+                    http_response.into_string()
+                ))
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Write for StreamingTransferWriter {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        // Write into memory buffer, and upload to GCS if we have accumulated
+        // enough content
+        self.buffer.extend_from_slice(buf);
+        if self.buffer.len() >= self.minimum_upload_chunk_size {
+            self.upload_chunk(false).map_err(|e| {
+                std::io::Error::new(std::io::ErrorKind::Other, Error::AnyhowError(e))
+            })?;
+        }
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+impl TransportWriter for StreamingTransferWriter {
+    fn complete_upload(&mut self) -> Result<()> {
+        while !self.buffer.is_empty() {
+            self.upload_chunk(true)?;
+        }
+        Ok(())
+    }
+
+    fn cancel_upload(&mut self) -> Result<()> {
+        // https://cloud.google.com/storage/docs/performing-resumable-uploads#cancel-upload
+        let http_response = ureq::delete(&self.upload_session_uri)
+            .set("Content-Length", "0")
+            // By default, ureq will wait forever to connect or read
+            .timeout_connect(10_000) // ten seconds
+            .timeout_read(10_000) // ten seconds
+            .call();
+        if http_response.status() != 499 {
```

Comment:
I am surprised GCS signals success with a 499! 🤷🏻 

Slight style nit: I would probably make this a final match statement:

```
match http_response.status() {
  499 => Err(..),
  _ => Ok(()),
}
```